### ISSUE TITLE
Redirection behaviour not consistent between Python2 & 3

### ISSUE THREAD

**Issue by BastiTee** [kennethreitz/requests#3949]
Hi there,
consider this minimal test:

```
from requests import exceptions, Session

url="http://ansgar-skoda.tumblr.com/post/96703389502"

try:
    s = Session()
    r = s.get(url)
except exceptions.TooManyRedirects as t:
    print('ERROR: {}'.format(t))
    r = s.get(url, allow_redirects=False)

print('========')
print(r.status_code)
print(r.headers)
```

Calling the link in a browser or via `curl -L` will follow the redirects and return a HTTP 200 after the redirection. That's what I expect.

The code above acts according to expectation when executed with Python2 but not with Python3. 
With Python3 you'll get stuck in a redirection loop that gets broken by the TooManyRedirection exception.

In both cases I installed requests 2.13.0 via pip.

I originally posted that on [Stackoverflow](http://stackoverflow.com/questions/43155127/how-to-correctly-setup-redirection-with-python-requests-for-some-corner-cases).

**Comment by nateprewitt** [comment]
Hey @BastiTee, thanks for taking the time to open this. The reason you're seeing this work in Python 2 vs Python 3 is the character encoding used. It also appears to be a dupe of #3888 but that puts more weight towards addressing this.

Since Python 2 has bytes as the default representation for strings, we're receiving the redirect url which ends with `gewÃ¤sch-und#_=_` as the already `latin-1` encoded byte sequence `gew\xc3\xa4sch-und#_=_`. This allows the redirect to succeed because that's the correct encoding for the address.

In Python 3, it stores the strings as `utf-8` by default, which when decoded back to bytes ends with the utf-8 equivalent byte sequence `gew\xc3\x83\xc2\xa4sch-und#_=_`. This is causing the redirect to continuously loop because the utf-8 encoding isn't correct.

For the time being, you should be able to manually encode the redirect headers containing non-ascii characters as `latin-1` bytes in Python 3. That's obviously not ideal, but seems to be the only thing we can suggest at present.

**Comment by Lukasa** [comment]
Yeah, I'm closing this as a dupe. The suggested fix on #3888 should be the direction we go.

**Comment by BastiTee** [comment]
Okay, thanks for suggesting a workaround. For so. else here's a workaround using the information above:
```
import requests
from requests import exceptions

url="http://ansgar-skoda.tumblr.com/post/96703389502"

try:
    r = requests.get(url)
except exceptions.TooManyRedirects as t:
    # see https://github.com/kennethreitz/requests/issues/3949
    r = requests.get(url, allow_redirects=False)
    latin1_location = r.headers['Location'].encode('latin1')
    r = requests.get(latin1_location)
    
print('========')
print(r.status_code)
print(r.headers)
```
